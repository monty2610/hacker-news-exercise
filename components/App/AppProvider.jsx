import React, { useReducer } from 'react';
import PropTypes from 'prop-types';

export const AppContext = React.createContext(null);

const reducer = (state, action) => {
  switch (action.type) {
    case 'FETCH_POST_SUCCESS':
      return {
        posts: [...state.posts, ...action.posts],
        page: action.page,
        fetchLatest: action.fetchLatest || false,
      };
    case 'UPDATE_POST':
      return {
        ...state,
        posts: action.posts,
      };
    case 'RESET':
      return {
        ...state,
        posts: [],
        page: 1,
      };
    default:
      return state;
  }
};

const AppProvider = ({ children, posts = [] }) => {
  const [state, dispatch] = useReducer(reducer, {
    posts,
    page: 1,
    fetchLatest: false,
  });

  return (
    <AppContext.Provider value={[state, dispatch]}>
      {children}
    </AppContext.Provider>
  );
};

AppProvider.propTypes = {
  children: PropTypes.node.isRequired,
  posts: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

export default AppProvider;
