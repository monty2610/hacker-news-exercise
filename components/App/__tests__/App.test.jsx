import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import AppProvider from '../AppProvider';
import AppComponent from '../App';

describe('App Component', () => {
  it('App component should render AppProvider', () => {
    const posts = [];
    const component = shallow(<AppComponent posts={posts} />);

    expect(component.is(AppProvider)).toBeTruthy();
  });

  it('App component should render correctly', () => {
    const posts = [];
    const tree = renderer.create(<AppComponent posts={posts} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
