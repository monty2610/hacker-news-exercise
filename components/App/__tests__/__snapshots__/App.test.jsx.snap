// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component App component should render correctly 1`] = `
Array [
  <nav
    className="jsx-1896995305"
  >
    <img
      alt="icon"
      className="jsx-1896995305 icon"
      src="/icons/y18.gif"
    />
    <button
      className="jsx-1896995305 active"
      onClick={[Function]}
      type="button"
    >
      top
    </button>
    <span
      className="jsx-1896995305"
    >
       | 
    </span>
    <button
      className="jsx-1896995305 "
      onClick={[Function]}
      type="button"
    >
      new
    </button>
  </nav>,
  <ul
    className="jsx-130854398"
  >
    <li
      className="jsx-130854398"
    >
      <button
        className="jsx-602828219"
        onClick={[Function]}
        type="button"
      >
        More
      </button>
    </li>
  </ul>,
]
`;
